timelock-agent/src/main/java/com/palantir/timelock/utils/KubernetesHostnames.java

```diff
@@ -0,0 +1,98 @@
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.palantir.logsafe.SafeArg;
+
+public final class KubernetesHostnames {
+
+    public static final KubernetesHostnames INSTANCE = new KubernetesHostnames(() -> {
+        try {
+            return InetAddress.getLocalHost().getCanonicalHostName();
+        } catch (UnknownHostException e) {
+            throw new RuntimeException(e);
+        }
+    });
+
+    private static final Logger log = LoggerFactory.getLogger(KubernetesHostnames.class);
+    private static final String POD_HOSTNAME_TEMPLATE = "%s-%d.%s.%s";
+
+    private final Supplier<String> currentHostnameSupplier;
+
+    /** The pattern for hostnames when running in a k8s statefulset. */
+    private static final Pattern POD_HOSTNAME_TEMPLATE_REGEX = Pattern.compile(
+            "(?<service>[a-z0-9\\-.]+)-(?<podId>\\d+)\\.\\k<service>"
+                    + "\\.(?<namespace>[a-z0-9\\-.]+)\\.svc\\.cluster\\.local");
+
+    @VisibleForTesting
+    KubernetesHostnames(Supplier<String> currentHostnameSupplier) {
+        this.currentHostnameSupplier = currentHostnameSupplier;
+    }
+
+    public String getCurrentHostname() {
+        Matcher matcher = getHostnameComponents();
+        return String.format(
+                POD_HOSTNAME_TEMPLATE,
+                matcher.group("service"),
+                Integer.valueOf(matcher.group("podId")),
+                matcher.group("service"),
+                matcher.group("namespace"));
+    }
+
+    public List<String> getClusterMembers(int expectedClusterSize) {
+        Matcher podTemplateMatcher = getHostnameComponents();
+
+        int ourPodId = Integer.parseInt(podTemplateMatcher.group("podId"));
+        Preconditions.checkArgument(ourPodId < expectedClusterSize,
+                "Current Pod ID %s indicates a cluster size greater than the expected %s.",
+                ourPodId,
+                expectedClusterSize);
+        String serviceName = podTemplateMatcher.group("service");
+        String namespace = podTemplateMatcher.group("namespace");
+        return IntStream.range(0, expectedClusterSize)
+                .mapToObj(podId -> getPodHostname(serviceName, podId, namespace))
+                .collect(Collectors.toList());
+    }
+
+    private String getPodHostname(String serviceName, int podId, String namespace) {
+        return String.format(POD_HOSTNAME_TEMPLATE, serviceName, podId, serviceName, namespace);
+    }
+
+    private Matcher getHostnameComponents() {
+        String canonicalHostName = currentHostnameSupplier.get();
+        log.info("Using hostname.", SafeArg.of("hostname", canonicalHostName));
+        Matcher matcher = POD_HOSTNAME_TEMPLATE_REGEX.matcher(canonicalHostName);
+        Preconditions.checkState(matcher.matches(), "Not running in a k8s stateful set.");
+        log.info("Hostname matcher results.", SafeArg.of("matcher", matcher));
```

Comment:
nit: if I'm not wrong, logging the matcher doesn't yield particularly helpful output. Not essential though - we can fix this later on.

Reply:
was mostly using it for initial debugging; going to remove for now